File: src/hab/utils/output.py
"""Output formatting utilities."""

from __future__ import annotations

import json
from datetime import UTC, datetime
from typing import Any

from rich.console import Console
from rich.table import Table
from rich.tree import Tree

console = Console()
error_console = Console(stderr=True)


def format_output(
    data: Any,
    *,
    text_mode: bool = False,
    success: bool = True,
    message: str | None = None,
) -> str:
    """Format output data.

    Args:
        data: Data to format.
        text_mode: Use human-readable text format instead of JSON.
        success: Whether the operation was successful.
        message: Optional message to include.

    Returns:
        Formatted output string.
    """
    if text_mode:
        return _format_text(data, message)

    output = {
        "success": success,
        "data": data,
        "metadata": {
            "timestamp": datetime.now(UTC).isoformat().replace("+00:00", "Z"),
        },
    }
    if message:
        output["message"] = message

    return json.dumps(output, indent=2, default=str)


def format_error(
    error: Exception | str,
    *,
    text_mode: bool = False,
    code: str = "GENERAL_ERROR",
    details: dict[str, Any] | None = None,
    suggestion: str | None = None,
) -> str:
    """Format error output.

    Args:
        error: Error exception or message.
        text_mode: Use human-readable text format instead of JSON.
        code: Error code.
        details: Additional error details.
        suggestion: Suggested action to fix the error.

    Returns:
        Formatted error string.
    """
    message = str(error)

    if text_mode:
        output = f"Error: {message}"
        if suggestion:
            output += f"\nSuggestion: {suggestion}"
        return output

    error_data: dict[str, Any] = {
        "code": code,
        "message": message,
    }
    if details:
        error_data["details"] = details
    if suggestion:
        error_data["details"] = error_data.get("details", {})
        error_data["details"]["suggestion"] = suggestion

    return json.dumps({"success": False, "error": error_data}, indent=2)


def _format_text(data: Any, message: str | None = None) -> str:
    """Format data as human-readable text.

    Args:
        data: Data to format.
        message: Optional message to include.

    Returns:
        Formatted text.
    """
    if message:
        return message

    if data is None:
        return "Done."

    if isinstance(data, str):
        return data

    if isinstance(data, bool):
        return "Yes" if data else "No"

    if isinstance(data, (int, float)):
        return str(data)

    if isinstance(data, list):
        return _format_list(data)

    if isinstance(data, dict):
        return _format_dict(data)

    return str(data)


def _format_list(data: list) -> str:
    """Format a list as text.

    Args:
        data: List to format.

    Returns:
        Formatted text.
    """
    if not data:
        return "No items."

    # If list of dicts with common keys, format as table
    if all(isinstance(item, dict) for item in data):
        return _format_dict_list(data)

    # Simple list
    lines = []
    for item in data:
        if isinstance(item, dict):
            # Extract meaningful identifier
            name = item.get("friendly_name") or item.get("name") or item.get("entity_id") or str(item)
            lines.append(f"  - {name}")
        else:
            lines.append(f"  - {item}")
    return "\n".join(lines)


def _format_dict_list(data: list[dict]) -> str:
    """Format a list of dicts as a table.

    Args:
        data: List of dicts to format.

    Returns:
        Formatted text.
    """
    if not data:
        return "No items."

    # Get common keys (limit to first 6 for readability)
    keys = list(data[0].keys())[:6]

    lines = []

    # Header
    header = " | ".join(str(k).replace("_", " ").title() for k in keys)
    lines.append(header)
    lines.append("-" * len(header))

    # Rows
    for item in data[:50]:  # Limit to 50 rows
        values = []
        for k in keys:
            v = item.get(k, "")
            if isinstance(v, (dict, list)):
                v = "..."
            values.append(str(v)[:30])
        lines.append(" | ".join(values))

    if len(data) > 50:
        lines.append(f"... and {len(data) - 50} more items")

    return "\n".join(lines)


def _format_dict(data: dict) -> str:
    """Format a dict as text.

    Args:
        data: Dict to format.

    Returns:
        Formatted text.
    """
    lines = []

    for key, value in data.items():
        key_label = str(key).replace("_", " ").title()

        if isinstance(value, dict):
            lines.append(f"{key_label}:")
            for k, v in value.items():
                lines.append(f"  {k}: {v}")
        elif isinstance(value, list):
            lines.append(f"{key_label}:")
            for item in value[:10]:
                if isinstance(item, dict):
                    lines.append(f"  - {item.get('name', item.get('entity_id', item))}")
                else:
                    lines.append(f"  - {item}")
            if len(value) > 10:
                lines.append(f"  ... and {len(value) - 10} more")
        else:
            lines.append(f"{key_label}: {value}")

    return "\n".join(lines)


def print_output(
    data: Any,
    *,
    text_mode: bool = False,
    success: bool = True,
    message: str | None = None,
) -> None:
    """Print formatted output to stdout.

    Args:
        data: Data to output.
        text_mode: Use human-readable text format.
        success: Whether the operation was successful.
        message: Optional message to include.
    """
    output = format_output(data, text_mode=text_mode, success=success, message=message)
    console.print(output, highlight=False)


def print_error(
    error: Exception | str,
    *,
    text_mode: bool = False,
    code: str = "GENERAL_ERROR",
    details: dict[str, Any] | None = None,
    suggestion: str | None = None,
) -> None:
    """Print formatted error to stderr.

    Args:
        error: Error exception or message.
        text_mode: Use human-readable text format.
        code: Error code.
        details: Additional error details.
        suggestion: Suggested action to fix the error.
    """
    output = format_error(
        error,
        text_mode=text_mode,
        code=code,
        details=details,
        suggestion=suggestion,
    )
    error_console.print(output, highlight=False)

File: client/output.go
package client

import (
	"encoding/json"
	"fmt"
	"strings"
	"time"
)

// Response represents the standard JSON output format
type Response struct {
	Success  bool                   `json:"success"`
	Data     interface{}            `json:"data,omitempty"`
	Message  string                 `json:"message,omitempty"`
	Error    *ErrorDetail           `json:"error,omitempty"`
	Metadata map[string]interface{} `json:"metadata,omitempty"`
}

// ErrorDetail contains error information
type ErrorDetail struct {
	Code    string                 `json:"code"`
	Message string                 `json:"message"`
	Details map[string]interface{} `json:"details,omitempty"`
}

// FormatOutput formats data for output
func FormatOutput(data interface{}, textMode bool, message string) string {
	if textMode {
		return formatText(data, message)
	}
	return formatJSON(data, true, message, nil)
}

// FormatSuccess formats a successful response
func FormatSuccess(data interface{}, message string) string {
	return formatJSON(data, true, message, nil)
}

// FormatError formats an error response
func FormatError(code string, msg string, details map[string]interface{}) string {
	return formatJSON(nil, false, "", &ErrorDetail{
		Code:    code,
		Message: msg,
		Details: details,
	})
}

// FormatErrorText formats an error for text output
func FormatErrorText(msg string, suggestion string) string {
	output := fmt.Sprintf("Error: %s", msg)
	if suggestion != "" {
		output += fmt.Sprintf("\nSuggestion: %s", suggestion)
	}
	return output
}

func formatJSON(data interface{}, success bool, message string, errDetail *ErrorDetail) string {
	resp := Response{
		Success: success,
		Data:    data,
		Message: message,
		Error:   errDetail,
		Metadata: map[string]interface{}{
			"timestamp": time.Now().UTC().Format(time.RFC3339),
		},
	}

	// Remove empty fields
	if message == "" {
		resp.Message = ""
	}
	if data == nil && success {
		resp.Data = nil
	}

	b, err := json.MarshalIndent(resp, "", "  ")
	if err != nil {
		return fmt.Sprintf(`{"success": false, "error": {"code": "MARSHAL_ERROR", "message": %q}}`, err.Error())
	}
	return string(b)
}

func formatText(data interface{}, message string) string {
	if message != "" {
		return message
	}

	if data == nil {
		return "Done."
	}

	switch v := data.(type) {
	case string:
		return v
	case bool:
		if v {
			return "Yes"
		}
		return "No"
	case int, int64, float64:
		return fmt.Sprintf("%v", v)
	case []interface{}:
		return formatList(v)
	case map[string]interface{}:
		return formatDict(v)
	default:
		// Try to marshal and re-parse as generic type
		b, err := json.Marshal(v)
		if err != nil {
			return fmt.Sprintf("%v", v)
		}
		var parsed interface{}
		if err := json.Unmarshal(b, &parsed); err != nil {
			return string(b)
		}
		return formatText(parsed, "")
	}
}

func formatList(data []interface{}) string {
	if len(data) == 0 {
		return "No items."
	}

	// Check if list of maps with common keys
	allMaps := true
	for _, item := range data {
		if _, ok := item.(map[string]interface{}); !ok {
			allMaps = false
			break
		}
	}

	if allMaps {
		return formatDictList(data)
	}

	// Simple list
	var lines []string
	for _, item := range data {
		if m, ok := item.(map[string]interface{}); ok {
			name := getDisplayName(m)
			lines = append(lines, fmt.Sprintf("  - %s", name))
		} else {
			lines = append(lines, fmt.Sprintf("  - %v", item))
		}
	}
	return strings.Join(lines, "\n")
}

func formatDictList(data []interface{}) string {
	if len(data) == 0 {
		return "No items."
	}

	// Get keys from first item
	first, ok := data[0].(map[string]interface{})
	if !ok {
		return "No items."
	}

	var keys []string
	for k := range first {
		keys = append(keys, k)
		if len(keys) >= 6 { // Limit columns
			break
		}
	}

	var lines []string

	// Header
	var headerParts []string
	for _, k := range keys {
		headerParts = append(headerParts, formatKey(k))
	}
	header := strings.Join(headerParts, " | ")
	lines = append(lines, header)
	lines = append(lines, strings.Repeat("-", len(header)))

	// Rows
	maxRows := 50
	for i, item := range data {
		if i >= maxRows {
			break
		}
		m, ok := item.(map[string]interface{})
		if !ok {
			continue
		}
		var values []string
		for _, k := range keys {
			v := m[k]
			str := formatValue(v)
			if len(str) > 30 {
				str = str[:27] + "..."
			}
			values = append(values, str)
		}
		lines = append(lines, strings.Join(values, " | "))
	}

	if len(data) > maxRows {
		lines = append(lines, fmt.Sprintf("... and %d more items", len(data)-maxRows))
	}

	return strings.Join(lines, "\n")
}

func formatDict(data map[string]interface{}) string {
	var lines []string

	for key, value := range data {
		keyLabel := formatKey(key)

		switch v := value.(type) {
		case map[string]interface{}:
			lines = append(lines, fmt.Sprintf("%s:", keyLabel))
			for k, val := range v {
				lines = append(lines, fmt.Sprintf("  %s: %s", k, formatValue(val)))
			}
		case []interface{}:
			lines = append(lines, fmt.Sprintf("%s:", keyLabel))
			maxItems := 10
			for i, item := range v {
				if i >= maxItems {
					lines = append(lines, fmt.Sprintf("  ... and %d more", len(v)-maxItems))
					break
				}
				if m, ok := item.(map[string]interface{}); ok {
					lines = append(lines, fmt.Sprintf("  - %s", getDisplayName(m)))
				} else {
					lines = append(lines, fmt.Sprintf("  - %v", item))
				}
			}
		default:
			lines = append(lines, fmt.Sprintf("%s: %s", keyLabel, formatValue(value)))
		}
	}

	return strings.Join(lines, "\n")
}

func formatKey(key string) string {
	// Convert snake_case to Title Case
	parts := strings.Split(key, "_")
	for i, p := range parts {
		if len(p) > 0 {
			parts[i] = strings.ToUpper(p[:1]) + p[1:]
		}
	}
	return strings.Join(parts, " ")
}

func formatValue(v interface{}) string {
	switch val := v.(type) {
	case nil:
		return ""
	case string:
		return val
	case bool:
		if val {
			return "Yes"
		}
		return "No"
	case map[string]interface{}, []interface{}:
		return "..."
	default:
		return fmt.Sprintf("%v", val)
	}
}

func getDisplayName(m map[string]interface{}) string {
	// Try common name fields
	for _, key := range []string{"friendly_name", "name", "entity_id", "id"} {
		if v, ok := m[key]; ok {
			return fmt.Sprintf("%v", v)
		}
	}
	return fmt.Sprintf("%v", m)
}

// PrintOutput prints formatted output to stdout
func PrintOutput(data interface{}, textMode bool, message string) {
	output := FormatOutput(data, textMode, message)
	fmt.Println(output)
}

// PrintSuccess prints a successful response
func PrintSuccess(data interface{}, textMode bool, message string) {
	if textMode {
		fmt.Println(formatText(data, message))
	} else {
		fmt.Println(FormatSuccess(data, message))
	}
}
